File: lib/devices/core/registry/device_classification.dart
import '../../../domain/model/ble_sensor_models.dart';

/// How the app should treat a discovered device: which integration owns it
/// (`null` = a plain live sensor, owned by no file-sync integration) and as what
/// [BleDeviceKind]. The authoritative mapping from a scanned device to how it is
/// registered and driven.
class DeviceClassification {
  const DeviceClassification({this.integration, required this.kind});

  final DeviceIntegration? integration;
  final BleDeviceKind kind;

  /// The default: a live BLE sensor belonging to no file-sync integration.
  static const DeviceClassification sensor =
      DeviceClassification(kind: BleDeviceKind.sensor);
}

/// One integration's verdict on a scanned device — its [DeviceClassification],
/// or `null` when the device is not its own. Each integration (Garmin, WearOS,
/// …) supplies one; the scanner/onboarding asks them all, so no generic code
/// names a protocol. Mirrors the advertisement-shaped [DeviceScanClassifier]
/// already in `core/ble`, but decides the whole (integration, kind) mapping.
abstract interface class DeviceClassifier {
  DeviceClassification? classify(BleDiscoveredDevice device);
}

/// Maps [device] to how the app should treat it: the first [classifiers] verdict
/// that claims it, else a plain [DeviceClassification.sensor]. Order matters —
/// pass the stronger signal first (Garmin's member service beats a name match).
DeviceClassification classifyDevice(
  BleDiscoveredDevice device,
  Iterable<DeviceClassifier> classifiers,
) {
  for (final classifier in classifiers) {
    final verdict = classifier.classify(device);
    if (verdict != null) return verdict;
  }
  return DeviceClassification.sensor;
}

File: app/src/main/kotlin/tech/mmarca/openvitals/devices/core/DeviceClassification.kt
package tech.mmarca.openvitals.devices.core

import tech.mmarca.openvitals.domain.model.BleDeviceKind
import tech.mmarca.openvitals.domain.model.BleDiscoveredDevice
import tech.mmarca.openvitals.domain.model.DeviceIntegration

/** How the app treats a discovered device: the owning integration (null for a plain sensor) and its kind. */
data class DeviceClassification(
    val integration: DeviceIntegration? = null,
    val kind: BleDeviceKind,
) {
    companion object {
        /** The default: a live BLE sensor belonging to no file-sync integration. */
        val SENSOR = DeviceClassification(kind = BleDeviceKind.SENSOR)
    }
}

/** One integration's verdict on a scanned device, or null when it is not its own. */
fun interface DeviceClassifier {
    fun classify(device: BleDiscoveredDevice): DeviceClassification?
}

/** The first [classifiers] verdict that claims [device], else a plain sensor. Stronger signals first. */
fun classifyDevice(
    device: BleDiscoveredDevice,
    classifiers: Iterable<DeviceClassifier>,
): DeviceClassification {
    for (classifier in classifiers) {
        val verdict = classifier.classify(device)
        if (verdict != null) return verdict
    }
    return DeviceClassification.SENSOR
}
